initial framework for database reconcile

Madhu-1 opened the pull request:
Added code for the below functionality

* Connect to the database
* Create a Notification Handler to receive notification
* Read all the entries from the database table and send them to Formen.

leelavg commented:
- Could you pls remove vendor content from the PR?

Madhu-1 commented:
> * Could you pls remove vendor content from the PR?

can you please explain why it need to be removed?

nb-ohad commented on config/config.yaml:
```diff
@@ -0,0 +1,7 @@
+db_user: postgres
+db_password: postgres
+db_name: mydatabase
+table_name: provider_service_table
```
```suggestion
db: 
     host: 192.168.178.20
     port: 5432
     user: postgres
     password: postgres
     name: mydatabase
     providers_table_name: provider_service_table
```

Madhu-1 committed:
initial framework for database reconcile

Added code for the below functionality

Connect to the database
Create a Notification Handler to receive notification
Read all the entries from the database table and
send them to Formen.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
vendor: vendor changes

vendor changes for the skeleton

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
 
 func main() {
+	var logger *zap.Logger
+	var err error
```
Polluting the function scope with a generic error variable has some cons. 
Mainly the inability to tie a memory address to a specific error when you do low-level debugging. 
Let's try to have local error variables for each return error.

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+	}
+	ctx := context.Background()
+	// parse configuration from yaml file
+	conf, err := db.GetConfig(logger, "config/config.yaml")
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
```
An example of local error variables. Please apply to all relevant statements below.
```suggestion
	if conf, err := db.GetConfig(logger, "config/config.yaml"); err != nil {
		logger.Fatal(err.Error())
	}
```

Madhu-1 replied:
we cannot access the conf as it can only be accessed inside the condition

nb-ohad replied:
You can predefine it in the outer scope. The issue you see here is shadowing

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+	}
+	ctx := context.Background()
+	// parse configuration from yaml file
+	conf, err := db.GetConfig(logger, "config/config.yaml")
```
If the file is hardcoded into the binary, and is parsed into a go struct, why can't we just have the config as go struct from the start? So instead of `config.yaml` we will have a `config.go` file

Madhu-1 replied:
i was thinking of mounting this as a file during the deployment not to have hardcoded into the binary and have it as configurable

nb-ohad replied:
Mount it from where?

Madhu-1 replied:
During the installation/deployment time mount it as a file and the user/admin can tune it based on the requirements.

nb-ohad replied:
We are not running this on K8s, we are running this on IBM code engine. I am not sure if your approach is valid in that case. Maybe a bunch of env vars will work better. Please research and provide an answer

Madhu-1 replied:
am not sure its a good idea to pass the configuration such as password in env. let me check how IBM code engine works

Madhu-1 replied:
we can create secret and mount it to a directory when running with code engine https://cloud.ibm.com/docs/codeengine?topic=codeengine-secretcm-reference-mountedfiles

nb-ohad replied:
For security concerns, all passwords need to be grabbed from a KMS link service. 
That also applies to the DB service - we don't have to focus on it in this PR. But eventually, we will need to get there.

For the rest of the config, if you feel that this should be something that we mount from the outside then OK. But if we are to do so, please aggregate all of the need configurations into a single file. 

For example, I saw that you defined a DEVELOPMENT envvar. Lets move that into the config as well

Madhu-1 replied:
yeah that's a good idea, updated the code to have a generic section in config where we can put generic configuration, let me know if you suggest to name it differently

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,127 @@
+	if cfg.Workers == 0 {
+		cfg.Workers = 5
+		logger.Warn("workers is not set, defaulting to", zap.Int("workers", cfg.Workers))
+	}
```
Let's make sure that specifying the workers count is mandatory.

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,127 @@
+	Workers    int    `yaml:"workers"`
+}
+
+func GetConfig(logger *zap.Logger, filePath string) (*Config, error) {
```
This is more of a ValidateConfig function then a GetConfig function. 
Also handles concerns that our outside of DB configuration (e.g. worker count)

Madhu-1 replied:
do we need to have two functions one to get config and one to validate it?

nb-ohad replied:
We should not have a function to get the config at all. It is embedded in the container.

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,127 @@
+
+func (c *Config) GetConnectionString() string {
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", c.DBHost, c.DBPort, c.DBUser, c.DBPassword, c.DBName)
+}
```
```suggestion
func GetConnectionString(c *Config) string {
	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", c.DBHost, c.DBPort, c.DBUser, c.DBPassword, c.DBName)
}
```

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,109 @@
+			go func() { toWorkers <- req }()
+		}
+
+		logger.Info("Starting workers", zap.Int("count", workerCount))
```
Why are we using machine readable messages and not human readable messages
```Starting workers { count: # }```
vs
``` Starting # workers```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+type Postgres struct {
+	conf *Config
```
Why do we need to piggyback the config after we created the DB instance?

Madhu-1 replied:
removed it now

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,127 @@
+	MetaData    sql.NullString
+	Spec        sql.NullString
+	Status      sql.NullString
+}
```
I think our entities should be DB agnostic and live in the reconcile package, or even a types package, but not the DB package

nb-ohad commented on .golangci.yaml:
```diff
@@ -0,0 +1,18 @@
+# This file contains all available configuration options
+# with their default values.
+# options for analysis running
+run:
+  # timeout for analysis, e.g. 30s, 5m, default is 1m
+  timeout: 2m
+linters:
+  disable-all: true
+  enable:
+    - deadcode
+    - errcheck
+    - gosimple
+    - ineffassign
+    - staticcheck
+    - structcheck
+    - unused
+    - varcheck
+    - revive
```
Is this our common/the common configuration?

Madhu-1 replied:
For now, this is the minimal configuration, we can enable a few extra checks

nb-ohad replied:
My question is where did you get that list from?

Madhu-1 replied:
i got this as reference from here https://github.com/red-hat-storage/ocs-client-operator/blob/main/.golangci.yaml

nb-ohad replied:
Let's try to comply with other components of the same project and not repos from other projects. A good place to look at will be https://github.com/red-hat-storage/managed-fusion-agent

Madhu-1 replied:
Sure, removed the check in the makefile

nb-ohad commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+# Build stage 2
+
+FROM registry.access.redhat.com/ubi9/ubi-minimal
```
Is this supposed to be for upstream only? if not I don't think we should pull from redhat registry

Madhu-1 replied:
replaced with static:nonroot image

nb-ohad commented on Makefile:
```diff
@@ -1,3 +1,78 @@
+export GO111MODULE=on
+# Enable GOPROXY. This speeds up a lot of vendoring operations.
+export GOPROXY=https://proxy.golang.org
+# Export GOROOT. Required for OPERATOR_SDK to work correctly for generate commands.
+export GOROOT=$(shell go env GOROOT)
```
Do we need this? we are not using OPERATOR_SDK or anything similar

Madhu-1 replied:
not required, removed it

nb-ohad commented on Makefile:
```diff
@@ -1,3 +1,78 @@
+
+GO_LINT_IMG_LOCATION ?= golangci/golangci-lint
+GO_LINT_IMG_TAG ?= v1.45.2
+GO_LINT_IMG ?= $(GO_LINT_IMG_LOCATION):$(GO_LINT_IMG_TAG)
```
Can we have all of the variables defined at the top of the makefile before any make target? 
It is a best practice we are using in all of our projects.

I am not just talking about these 3 variables but for all variables in the file

nb-ohad commented on Makefile:
```diff
@@ -1,3 +1,78 @@
+	go vet ./...
+
+.PHONY: lint
+lint: ## Run golangci-lint against code.
+	docker run --rm -v $(PROJECT_DIR):/app:Z -w /app $(GO_LINT_IMG) golangci-lint run ./... -v
```
Please note that we have 3 lint tools (see the github action of the repo). Can we run all of them here?

Madhu-1 replied:
AFAIC we have 2 linter, golangci-lint and yaml lint, will send follow up PR to add yaml linter as well.

nb-ohad replied:
Ok

nb-ohad commented on Makefile:
```diff
@@ -1,3 +1,78 @@
+
+.PHONY: container-build
+container-build: ## Build container image with the managed-fusion-fleet-reconciler.
+	docker build -t ${IMG} .
```
What does `-t` means?
Also if we are hardcoding `docker` here, can we please call the target `docker-build` instead of `container-build`?
This will also align with the make target in our other repos

nb-ohad commented on Makefile:
```diff
@@ -1,3 +1,78 @@
+	docker build -t ${IMG} .
+
+.PHONY: container-push
+container-push: ## Push container image with the managed-fusion-fleet-reconciler.
```
Same suggestion as last comment, change to `docker-push`

nb-ohad commented on README.md:
```diff
@@ -1 +1 @@
-# managed-fusion-fleet-reconciler
\ No newline at end of file
+# managed-fusion-fleet-reconciler
```
What exactly is the change here? if there is none can we drop the change

Madhu-1 replied:
a new line was added which was missing

nb-ohad replied:
ok

nb-ohad commented on main.go:
You have a DB package that offers almost no high-level abstraction on top of the actual DB work, while this main package is implementing a lot of db logic. 

It will make more sense to move the majority of the steps into the DB package and have a high-level operation exported to set up the DB connection

Madhu-1 replied:
moved most of the logic to setup the notification into db

nb-ohad commented on pkg/reconciler/reconcile.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+func Reconcile(logger *zap.Logger, dbInstance *db.Postgres, req forman.Request) forman.Result {
+	ctx := context.TODO()
```
Let's use `context.Background()` and not `context.TODO()`. Or even better provide it from the outside

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+	}(providers)
+
+	for {
+		err = pgConn.WaitForNotification(context.Background())
```
Is there a reason to provide a new context here?

Madhu-1 replied:
reused context

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+		}
+	}(providers)
+
+	for {
```
Why do we need a tight loop here?

Madhu-1 replied:
This is required to read the notifications

nb-ohad replied:
Can you please elaborate?

Madhu-1 replied:
https://github.com/jackc/pgx/blob/master/doc.go#L160-L172, WaitForNotification will return once the notification is received so we need to call it in loop to handle the all the notifications

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+		logger.Fatal(err.Error())
+	}
+
+	go func(providers []*db.ProviderCluster) {
+		for _, p := range providers {
+			req := forman.Request{}
+			req.Name = p.ClusterID
+			reqChan <- req
+		}
+	}(providers)
```
This will queue a reconcile request for every provider, and all at the same time.
We should only queue for providers who had an unprocessed state change or at least not in a stable state

Madhu-1 replied:
Agree, will take care of it in follow up as need to experiment with json, will add a check for hash to compare current spec and desired spec

nb-ohad replied:
The compare is only relevant if the state is stable. If it is not we need to requeue anyways

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,94 @@
+	}
+	defer pgConn.Close(ctx)
+
+	// Listen for notifications
+	_, err = pgConn.Exec(ctx, "LISTEN provider_cluster;").ReadAll()
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
```
Not sure what this is

Madhu-1 replied:
This is a subscription for the notifications

nb-ohad replied:
In that case, where is `provider_cluster` defined?

Madhu-1 replied:
This is defined in the postgresql trigger what need to be setup in the database, for now we are not creating triggers not tables in the code, this might go away once we decided on that

nb-ohad replied:
so if we provide a table name in the config shouldn't we also provide the trigger name there? 
Or even use the table name as the trigger name

Madhu-1 replied:
updated code to use table name as the trigger name

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,109 @@
+	alarmClock *time.Timer
+}
+
+func GoForman(
+	logger *zap.Logger,
+	dbInstance *db.Postgres,
+	workerCount int,
+	reconcile func(*zap.Logger, *db.Postgres, Request) Result,
```
It is a bad idea to couple the Forman with the DB instance or have the reconcile function signature have db-related things. 

I might see a case for Logger here, but even that is a stretch

Madhu-1 replied:
db is a concurrent safe operation instance, do you want it to be removed for now?

nb-ohad replied:
I am saying that we should not introduce these types into the formal package. You can have additional information sent to the actual reconcile function by creating a closure in main

Madhu-1 replied:
removed the db parameter

nb-ohad commented on config/config.yaml:
```diff
@@ -0,0 +1,12 @@
+  name: mydatabase
+  providers_table_name: provider_service_table
+worker:
+  count: 5
```
```suggestion
reconcile:
    concurrency: 5
```

nb-ohad commented on config/config.yaml:
```diff
@@ -0,0 +1,12 @@
+  user: postgres
+  password: postgres
+  name: mydatabase
+  providers_table_name: provider_service_table
```
Why snake case?
```suggestion
  providersTableName: provider_service_table
```

nb-ohad commented on config/config.yaml:
```diff
@@ -0,0 +1,12 @@
+worker:
+  count: 5
+general:
+  development: true
```
Let's make sure we can run under different profiles, at least: 
development, testing, staging, production

```suggestion
runtimeEnvironment: 
  profile: "development"
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+		log.Fatal(err)
+	}
+
+	logger, err := zap.NewProduction()
+	if conf.General.Development {
+		logger, err = zap.NewDevelopment()
+	}
```
See comments about the config structure. This will need to change to a switch statement

Madhu-1 replied:
i think not required anymore as we use different logger

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	ctx := context.Background()
```
Can we make this line the first line of the main?

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+		logger.Fatal(err.Error())
+	}
+
+	if err := dbInstance.Ping(ctx); err != nil {
+		logger.Fatal(err.Error())
+	}
```
Why is that a different step? can't we validate a ping as part of instance creation?

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+
+	reqChan := forman.GoForman(logger, conf.Worker.Count, reconciler.Reconcile)
+
+	pgConn, err := dbInstance.SetupNotification(ctx, reqChan)
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
```
Why do we need a different step for setuping the notifications? can't we roll that into the instance creation or the listening part?

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+		logger.Fatal(err.Error())
+	}
+
+	go func(providers []*types.ProviderCluster) {
+		for _, p := range providers {
+			req := types.Request{}
+			req.Name = p.ClusterID
+			reqChan <- req
+		}
+	}(providers)
```
You don't need to send the `providers` into the function, just use it as part of a closure

Madhu-1 replied:
slightly modified the code to load one by one and send it to forman in go routine

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,67 @@
+		}
+	}(providers)
+
+	for {
+		if err := pgConn.WaitForNotification(ctx); err != nil {
+			logger.Fatal(err.Error())
+		}
+	}
```
Why are we implementing this here (in main). This looks like a db concern and I would like to see something like
```
dbInstance.ListenForNotifications(ctx) 
```
here

nb-ohad commented on pkg/config/config.go:
```diff
@@ -0,0 +1,58 @@
+	"gopkg.in/yaml.v2"
+)
+
+type Worker struct {
+	Count int `yaml:"count"`
+}
+
+type General struct {
+	Development bool `yaml:"development"`
+}
+type Config struct {
+	DB      db.DB   `yaml:"db"`
+	Worker  Worker  `yaml:"worker"`
+	General General `yaml:"general"`
+}
```
As we are talking about config a better approach might be to have a more compound type definition.
Also, please define all parts of the config in this package (db.Db) 

```suggestion
type Worker 

type General 
type Config struct {
	DB      struct {
	    .....
	}   `yaml:"db"`
	Worker  struct {
	        Count int `yaml:"count"`
        }  `yaml:"worker"`
	General struct {
	        Development bool `yaml:"development"`
        }`yaml:"general"`
}
```

Madhu-1 replied:
Implement this but created a different struct for accessing it and for easy maintaince.

nb-ohad commented on pkg/config/config.go:
```diff
@@ -0,0 +1,58 @@
+	General General `yaml:"general"`
+}
+
+func GetConfig(filePath string) (*Config, error) {
```
Use a more descriptive and complete name:
```suggestion
func LoadAndValidateConfig(filePath string) (*Config, error) {
```

nb-ohad commented on pkg/config/config.go:
```diff
@@ -0,0 +1,58 @@
+		return nil, err
+	}
+	if cfg.DB.Host == "" {
+		return nil, fmt.Errorf("db:host is not set")
```
Paths are donated with `.` not `:`, let try to comply with that 
```suggestion
		return nil, fmt.Errorf("config.db.host is not set")
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	Password  string `yaml:"password"`
+	Name      string `yaml:"name"`
+	TableName string `yaml:"providers_table_name"`
+}
```
Should move into the config pacakge

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	TableName string `yaml:"providers_table_name"`
+}
+
+func getConnectionString(db *DB) string {
```
```suggestion
func getConnectionString(config *config.DB) string {
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", db.Host, db.Port, db.User, db.Password, db.Name)
+}
+
+type Postgres struct {
+	db   *DB
+	pool *pgxpool.Pool
+}
```
The entire DB can be represented as an alias to pgxpool.Pool
```
type Database pgxpool.Pool 
```

Or just embed it if you think that is better
```
type Database {
    *pgxpool.Pool
}
```

leelavg replied:
- I feel similar too based on usage of `Postgres.db` in this pkg
- Apart from generating connection string we are only using TableName, however entire DB may not exist for only performing on this single table and as such I believe it's better not to have `db` in this struct

Madhu-1 replied:
updated struct to storage only the table name not the db connection

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	pool *pgxpool.Pool
+}
+
+func NewPGInstance(ctx context.Context, logger *zap.Logger, db *DB) (*Postgres, error) {
```
Why do we need to explicitly specify that this is a PG instance?
```suggestion
func NewClient(ctx context.Context, logger *zap.Logger, db *DB) (*Postgres, error) {
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+
+func (pg *Postgres) Ping(ctx context.Context) error {
+	return pg.pool.Ping(ctx)
+}
```
With the current use case there is no need to expose the ping functionality

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	}
+
+	return pgConn, err
+}
```
Why does the setup part need to be separate from the listening part?

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	}
+	defer rows.Close()
+
+	var cluster []*types.ProviderCluster
+	for rows.Next() {
+		var c types.ProviderCluster
+		err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status)
+		if err != nil {
+			return nil, err
+		}
+		cluster = append(cluster, &c)
+	}
```
As discussed in the last review, we do not want to load all providers into memory at the same time

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,99 @@
+
+		logger.Info("Starting", zap.Int("workers", workerCount))
+		for i := 0; i < workerCount; i++ {
+			workers.GoWorker(toWorkers, results, func(req types.Request) types.Result {
+				return reconcile(logger, req)
+			})
```
No need for that, we can enhance the worker API to include a logger as well. Mainly as it is a standard

Madhu-1 replied:
Sorry didnt get the suggestion

nb-ohad commented on pkg/reconciler/reconcile.go:
```diff
@@ -0,0 +1,12 @@
+	"go.uber.org/zap"
+)
+
+func Reconcile(logger *zap.Logger, req types.Request) types.Result {
```
You need to accept the Database Client here

nb-ohad commented on pkg/types/types.go:
```diff
@@ -0,0 +1,24 @@
+	ClusterID   string
+	AccountID   string
+	SatelliteID string
+	MetaData    sql.NullString
+	Spec        sql.NullString
+	Status      sql.NullString
```
As mentioned in the previous review, the types should not use package proprietary types. Also we are expecting sub-structures here and not strings

Madhu-1 replied:
Yes am aware of it as it was not clear how it looks like I kept it as String and will change it as we implement it.

nb-ohad replied:
Just put empty structs

nb-ohad commented on pkg/types/types.go:
```diff
@@ -0,0 +1,24 @@
+	Status      sql.NullString
+}
+
+type Request struct {
+	Name string `json:"name"`
+}
+
+type Result struct {
+	Requeue bool
+	After   time.Duration
+}
```
These types are part of the formen interface. They are not data types.
Lets keep them in the Fromen package

Madhu-1 replied:
i kept it here to avoid cycle import problem, this is also required in db folder to send the notifications and formen is also importing db package

nb-ohad replied:
That should not be a problem the DB package should not relay on these at all

Madhu-1 replied:
db also need to setup the notification as you suggested and for that we need this Request struct to be imported.

nb-ohad commented:
When passing a logger around (for example in func definitions), please use the `logger.Logger` interface and not the Zap concrete implementation  (see "github.com/go-logr/logr")

leelavg commented:
> can you please explain why it need to be removed?
- I feel the guarantees that vendor provides isn't required to be carried until we hit an issue for them to be useful
- As an example, vendor is helpful if a repo was taken down or intermittently unreachable, in the former case we (at-least me) don't want our builds to pass silently
- Nevertheless, I simply wanted it not to be part of this specific PR as that makes the UI slow, there are pros & cons for using(/not) and that decision isn't upto me

leelavg commented on Makefile:
```diff
@@ -1,3 +1,69 @@
+IMAGE_REGISTRY ?= quay.io
+REGISTRY_NAMESPACE ?= ocs-dev
```
- can we not have this var set and let it fail when not supplied?
- I'm not sure of all other vars and at the same time this mostly doesn't fit as a default destination

Madhu-1 replied:
What i have seen in most places is that we make default in the Makefile, if this is the standard followed in managed service I can switch to that model.

leelavg replied:
ack, what I only meant was `ocs-dev` is a bad default, since I don't have a good one, asked it to be empty.

nb-ohad replied:
I agree with @leelavg! why does the FaaS toolchain have an OCS default?

Madhu-1 replied:
@nb-ohad am not sure what do we have for Faas, I can update the same here, this is just a place holder

nb-ohad replied:
We don't! keep it empty

leelavg commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+COPY . .
+
+RUN GOOS=linux GOARCH=amd64 go build -tags 'netgo osusergo' -o ./bin/managed-fusion-fleet-reconciler main.go
```
1. could you pls mention what these build tags enables/disables in the tool chain?
2. do we have `CGO` as a dependency, if not, can that be switched off or else I feel the exe will fail to run in distroless?

Madhu-1 replied:
This is another way to build static libraries, changing it to CGO as most here are familiar with it

leelavg replied:
ack, as long as those tags does what env var achieves, I don't have any preference.

leelavg commented on Makefile:
```diff
@@ -1,3 +1,69 @@
+# Export GO111MODULE=on to enable project to be built from within GOPATH/src
+export GO111MODULE=on
+# Enable GOPROXY. This speeds up a lot of vendoring operations.
+export GOPROXY=https://proxy.golang.org
```
- shouldn't we be having `,direct` as well, I don't remember not having `direct` would still pull the non-proxy dependencies or not?

Madhu-1 replied:
yes correct, but removing it now lets use whats set as default in system

leelavg replied:
ack, so in Make recipes even if GOPROXY doesn't have `direct` go would still correctly pull the deps?

leelavg commented on main.go:
```diff
@@ -1,5 +1,67 @@
 
 func main() {
+	// parse configuration from yaml file
+	conf, err := config.GetConfig("config/config.yaml")
```
- will we be making sure that this gets mounted relative to the exe?
- pardon my knowledge gap wrt code engine, why not remove the dependency here and supply the config path from env?
- maybe I'm thinking too much, in addition to above I feel the default to be a failure, like, move `config/config.yaml` to `example/config.yaml` and add `config/` to `{.git,.docker}ignore`

Madhu-1 replied:
i dont see the reason to make it configurable as we decide how to deploy it. If you feel it needs to be changed, please let me know. sounds good to add it do dockerignore not gitignore

leelavg replied:
> make it configurable
- if we can always ensure this is mounted relative to exe I don't have any issue

> add it do dockerignore not gitignore
- if devs tries to update the config to prod creds I thought it's better to have this file ignored rather than committed by mistake

nb-ohad replied:
Moving this to be provided by the environment makes sense. The repo itself has no notion of deployment and as such of deployment location. I will vote to have this provided by the environment via and env var (CONFIG_FILE)

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	}
+	config.OnNotification = func(c *pgconn.PgConn, n *pgconn.Notification) {
+		req := types.Request{}
+		if err := json.Unmarshal([]byte(n.Payload), &req); err != nil {
```
- what does this `Payload` usually contain?

Madhu-1 replied:
it contains the json {`name` : `ID`}

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+	return pgConn, err
+}
+
+func (pg *Postgres) GetProviderCluster(ctx context.Context, clusterID string) (*types.ProviderCluster, error) {
```
- is this func unused? if yes, can this added when required?

Madhu-1 replied:
i added it thinking this is required as this is the first thing worker will do when receive a task, if you think its not required , removing it now

nb-ohad replied:
It is required. As of the comment from my review, the skeleton for reconciliation should already contain a to load the provider record

leelavg replied:
> as this is the first thing worker
- ack, with updated code, it's more evident.

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+
+	var cluster []*types.ProviderCluster
+	for rows.Next() {
+		var c types.ProviderCluster
+		err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status)
```
- if we have db schema and corresponding go types available, why can't we compile required queries ahead of time just to separate our code vs code which can be generated
- I suggest using this https://sqlc.dev/, we can discuss if required

Madhu-1 replied:
might be useful for future not for the skeleton

nb-ohad replied:
I am not sure I agree. Let's defer this decision for a later discussion

leelavg commented on main.go:
```diff
@@ -1,5 +1,67 @@
+
+	defer pgConn.Close(ctx)
+
+	providers, err := dbInstance.ListProviderClusters(ctx)
```
- is this a safety net against reboot? if yes, better to read the rows sequentially?

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,107 @@
+		cluster = append(cluster, &c)
+	}
+
+	return cluster, nil
```
- at the call site we are only using one field out of entire row, I feel it's better to store only that field in-case having all of them in memory is mandatory.

Madhu-1 replied:
Updated function not to store anything

nb-ohad replied:
I am not sure I agree, this loading process should have checks around provider phase, state, and desired state. As such you should load all of them.

Madhu-1 replied:
Yes that check is missing as we don't have defined state and phases, this function need to get updated later

leelavg commented on main.go:
```diff
@@ -1,5 +1,67 @@
+	}(providers)
+
+	for {
+		if err := pgConn.WaitForNotification(ctx); err != nil {
```
- referring to the [doc](https://github.com/jackc/pgx/blob/master/doc.go#L168) provided in another comment what are we doing with the notification and doesn't this method return two values?
- or we'll get only error due to already setting up notification earlier?

Madhu-1 replied:
its a wrapper over the pgconn.Notification here we are directly using pgconn and settling the Handler, if we go with this one we need to write our own handler to process the data

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,59 @@
 func main() {
+	//logger := logr.Logger{}
+	ctx := context.Background()
+	configFilePath := os.Getenv("CONFIG_FILE_PATH")
```
Can we make the env var more specific and remove the redundant "PATH" suffix?
```suggestion
	configFilePath := os.Getenv("FLEET_RECONCILER_CONFIG")
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,59 @@
+	ctx := context.Background()
+	configFilePath := os.Getenv("CONFIG_FILE_PATH")
+	if configFilePath == "" {
+		log.Fatal("CONFIG_FILE_PATH environment variable not set")
```
If you use the name of the env var more then once, please move it to a const at the top of the file
```
const configFileEnvVarName = "FLEET_RECONCILER_CONFIG"
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,59 @@
+		log.Fatal("CONFIG_FILE_PATH environment variable not set")
+	}
+	// parse configuration from yaml file
+	conf, err := config.LoadAndValidateConfig(configFilePath)
+	if err != nil {
+		log.Fatal(err)
+	}
```
Why do we need a new package for config? can we just implement this in `config.go` on the main package?

Madhu-1 replied:
Do you suggest to pass the db details as input parameters not as struct to the DB Initialization?

Madhu-1 replied:
AFAIK we cannot have two files with package main, instead of a new package moved this functions to main.go itself

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,59 @@
+		log.Fatal(err)
+	}
+
+	dbInstance, err := db.NewClient(ctx, &conf.DB)
```
This is not an instance, it is more a client. Lets reflect that in the name
```suggestion
	dbClient, err := db.NewClient(ctx, &conf.DB)
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,59 @@
+
+	dbInstance, err := db.NewClient(ctx, &conf.DB)
+	if err != nil {
+		logger.Fatal(err.Error())
```
Because the defer is not set yet you need to close the client in case of an error 
```suggestion
		dbInstance.close()
		logger.Fatal(err.Error())
```

leelavg replied:
- I don't think so, there can be scenarios where `dbInstance` can be `nil` so we need to defer after ensuring no error, (unless I understood the ask differently)

nb-ohad replied:
So do this:
```suggestion
		if dbInstance != nil {
			dbInstance.close()
		}
		logger.Fatal(err.Error())
```

nb-ohad replied:
This is needed unless we have guarantees
 that `db.NewClient` either returns an instance or error (never both). This is not the case with the current revision of the code

Madhu-1 replied:
This is not required incase of error dbClient will always be nil, I don't think anyone will return both

nb-ohad replied:
👍

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,58 @@
+	}
+	defer dbInstance.Close()
+
+	reqChan := forman.GoForman(logger, dbInstance, conf.Reconcile.Concurrency, reconciler.Reconcile)
```
As mentioned in a previous comment. DB related structures should not be part of the Formen interface!

Madhu-1 replied:
i think there is some gap on this between us , I suggest having a meeting on what it looks like.

nb-ohad replied:
No need for a meeting, I just asked remove any knowledge of the DB interface from any packages that does not need to now about it:
```
reqChan := forman.GoForman(
    logger, 
    conf.Reconcile.Concurrency, 
    func(req foremam.Request) foremam.Result {
        reconciler.reconcile(logger, dbInstance, req)
    },
)
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,58 @@
+
+	reqChan := forman.GoForman(logger, dbInstance, conf.Reconcile.Concurrency, reconciler.Reconcile)
+
+	go func() {
+		if err := dbInstance.LoadExistingProviderClusters(ctx, reqChan); err != nil {
+			logger.Fatal(err.Error())
+		}
+	}()
```
Why are we running this as a go function?

leelavg replied:
- just so I understand this correctly, a send on a channel is happening in this func and forced you to run this concurrently?
- if that's the case, we need to change the name of the method

Madhu-1 replied:
if we dont run it in go routine we need to wait until all the work is assigned before listening for the notifications, we might end up missing updates on the DB.

Madhu-1 replied:
@leelavg what is your suggestion

leelavg replied:
One, I don't have any issue with implementation. It's just that I couldn't think of a better name for that function, `Load.. ` implies sync however it isn't.

Bottomline, pls give another thought if nothing strikes pls disregard my comment.

nb-ohad replied:
> if we dont run it in go routine we need to wait until all the work is assigned before listening for the notifications, we might end up missing updates on the DB.

I don't see that happening. 
We have 3 responsibilities that need to run in parallel:
1. listen to notifications -> should run as part of the main goroutine
2. accept work and distribute to workers -> Forman 
3. reconcile -> worker

So as of above, there is no need for a fourth goroutine

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,58 @@
+		}
+	}()
+
+	for {
+		if err := dbInstance.ListenForNotifications(ctx, &conf.DB, reqChan); err != nil {
+			logger.Fatal(err.Error())
+		}
+	}
 }
```
As stated before, the loop itself should be part of the Listen function not main!

nb-ohad commented on pkg/config/config.go:
```diff
@@ -0,0 +1,57 @@
+	"gopkg.in/yaml.v2"
+)
+
+type DB struct {
+	Host               string `yaml:"host"`
+	Port               int    `yaml:"port"`
+	User               string `yaml:"user"`
+	Password           string `yaml:"password"`
+	Name               string `yaml:"name"`
+	ProvidersTableName string `yaml:"providersTableName"`
+}
```
Why is that a separate structure and not a nested struct as part of the config struct as requested in the last review?

Madhu-1 replied:
kept it as nested struct

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+	"github.com/red-hat-storage/managed-fusion-fleet-reconciler/pkg/types"
+)
+
+func getConnectionString(db *config.DB) string {
```
Why are we coupling the DB module with the program's config structure?

Madhu-1 replied:
i moved this to config to return a string with connection details to connect to db

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", db.Host, db.Port, db.User, db.Password, db.Name)
+}
+
+type Database struct {
+	tableName string
+	pool      *pgxpool.Pool
+}
```
The database structure (or alias) is supposed to represent the database client. Is there a reason to attach a specific table name to it? what will happen when we have more than on table?

Madhu-1 replied:
if we dont store the table we need to pass the table name from the caller function, if you suggest doing it, I can update it.

nb-ohad replied:
I think you missed the point I was trying to make
```suggestion
type Database struct {
	pool      *pgxpool.Pool
	tables   map[string]string 
}
```

or even better, (as out client only work on specific tables and is not designed to be generic):
```suggestion
type Database struct {
	pool      *pgxpool.Pool
	tables   struct {
	    providers: string
	}
}
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+	pg.pool.Close()
+}
+
+var once sync.Once
```
Do we really need an alias for a single use?

Madhu-1 replied:
This is removed now.

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+
+var once sync.Once
+
+func (pg *Database) ListenForNotifications(ctx context.Context, db *config.DB, reqChan chan types.Request) error {
```
Why do we need to pass the configuration again? didn't we establish everything on the client when creating the instance?

Madhu-1 replied:
This is a connection for notifications, not the previous connection.

nb-ohad replied:
Does not make sense. Even if PostgreSQL needs the connection string again we already sent it to the DB module os there should be no need to resend it again.

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+
+var once sync.Once
+
+func (pg *Database) ListenForNotifications(ctx context.Context, db *config.DB, reqChan chan types.Request) error {
```
The signature is missing a callback for what happens when a notification is raised. It is a bad idea to couple this module with the notion of former requests.

Madhu-1 replied:
when the notificaiton is raised we are sending that to the reqChal. Do you suggest to pass the callback function as input?

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+			if err := json.Unmarshal([]byte(n.Payload), &req); err != nil {
+				return
+			}
+			reqChan <- req
```
This should be an invocation of a callback

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+		return err
+	}
+
+	if err := pgConn.WaitForNotification(ctx); err != nil {
+		pgConn.Close(ctx)
+		return err
+	}
```
@Madhu-1 as you stated in one of your prev comments. this need to run in an infinite loop

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var c types.ProviderCluster
+		if err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status); err != nil {
+			return err
+		}
+		reqChan <- types.Request{
+			Name: c.ClusterID,
+		}
+	}
```
I don't think that a push model (channels) is the right approach here. You want the caller to be responsible for the next read

nb-ohad replied:
I am closing this thread as I mentioned how I would like this to look like (with code suggestions) on a newer review

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,101 @@
+	alarmClock *time.Timer
+}
+
+func GoForman(
+	logger *zap.Logger,
+	client *db.Database,
+	workerCount int,
+	reconcile func(*zap.Logger, *db.Database, types.Request) types.Result,
```
Please see prev reviews for reasoning
```suggestion
func GoForman(
	logger *zap.Logger,
	workerCount int,
	reconcile func(*zap.Logger, *db.Database, types.Request) types.Result,
```

Madhu-1 replied:
if we dont pass the db.Database Here, how the foreman can call the below code

```go
logger.Info("Starting", zap.Int("workers", workerCount))
		for i := 0; i < workerCount; i++ {
			workers.GoWorker(toWorkers, results, func(req types.Request) types.Result {
				return reconcile(logger, client, req)
			})
		}
```

nb-ohad replied:
I am closing this thread as I mentioned how I would like this to look like (with code suggestions) on a newer review

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,101 @@
+
+		logger.Info("Starting", zap.Int("workers", workerCount))
+		for i := 0; i < workerCount; i++ {
+			workers.GoWorker(toWorkers, results, func(req types.Request) types.Result {
+				return reconcile(logger, client, req)
+			})
```
Please see prev review for reasoning 
```suggestion
			workers.GoWorker(toWorkers, results, reconcile)
```

Madhu-1 replied:
This suggest is not working, it needs more changes in the worker cod and the question i have is who will send the dbclient and logger instance here?

nb-ohad replied:
I am closing this thread as I mentioned how I would like this to look like (with code suggestions) on a newer review

nb-ohad commented on pkg/reconciler/reconcile.go:
```diff
@@ -0,0 +1,23 @@
+	"go.uber.org/zap"
+)
+
+func Reconcile(logger *zap.Logger, client *db.Database, req types.Request) types.Result {
```
Please pass the logger around via the logr.Logger interface (as requested in previous reviews)

Madhu-1 replied:
AFAIK The above is not possible because zap.Logger and logr.Logger are not compatible to each other as we do in reconciler in other project

nb-ohad replied:
Why? this is exactly the strategy operator framework is taking for generated operators:
https://github.com/red-hat-storage/managed-fusion-agent/blob/main/main.go#L103

Madhu-1 replied:
This is not possible as discussed we need to write wrapper over zap for this one, will take care of it in followup PR

nb-ohad replied:
Ok, as you already clarified offline. resolving the comment.

nb-ohad commented on pkg/types/types.go:
```diff
@@ -0,0 +1,31 @@
+	Status      Status
+}
+
+type Request struct {
+	Name string `json:"name"`
+}
+
+type Result struct {
+	Requeue bool
+	After   time.Duration
+}
```
Please Formen interface types back to the Formen package as requested in previous reviews

leelavg commented on pkg/workers/worker.go:
```diff
@@ -0,0 +1,32 @@
+type WorkHandler[Req, Res any] func(Req) Res
+
+func GoWorker[Req, Res any](
+	reqChan chan Req,
+	resChan chan utils.Pair[Req, Res],
```
- just a thought, will it make more sense to also specify the intention of these channels not only in their names, like
```suggestion
	reqChan <- chan Req,
	resChan chan <- utils.Pair[Req, Res],
```

nb-ohad replied:
This is new to me. Is this a way to force Go to only allow sends and not receives?

leelavg replied:
- `req <-chan`: forces `req` to be used only for `receive` in function body
- `res chan<-`: forces `res` to be used only for `send` in function body

nb-ohad commented on config/config.yaml:
Can we bump this file up into the project root?

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+		OutputPaths:      []string{"stdout"},
+		ErrorOutputPaths: []string{"stderr"},
+	}
+	logger, err := zapConfig.Build()
```
How did you come up with this configuration?

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+		log.Fatal(err)
+	}
+
+	connString := getConnectionString(conf)
```
For the sake of separation of concerns, and as requested in previous commits. please send the separate bits to this function:  
```suggestion
	connString := getConnectionString(conf.DB.Host, conf.DB.Port, conf.DB.User, conf.DB.Password, conf.DB.Name)
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+	}
+
+	connString := getConnectionString(conf)
+	dbClient, err := db.NewClient(ctx, connString, conf.DB.ProvidersTableName)
```
Why does the client need to know the provider's table name? when we have more than one table are you intending to send the entire list to the client?

Madhu-1 replied:
Yes that was the plan, if you think otherwise and the callers need to send the table name (as the functions are not generic enough to work with all tables) I can do it as well.

nb-ohad replied:
Ok

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+	}
+	defer dbClient.Close()
+
+	reqChan := forman.GoForman(logger, dbClient, conf.Reconcile.Concurrency, reconciler.Reconcile)
```
I see Forman is still aware if  DB types. Please remove this, if you don't know how refer to the following suggestion
 ```suggestion
	reqChan := forman.GoForman(
		logger,
		conf.Reconcile.Concurrency, 
		func (req forman.Request) froman.Result {
			return reconciler.Reconcile(logger, dbClient, req)
		},
	)
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+	Reconcile struct {
+		Concurrency int `yaml:"concurrency"`
+	} `yaml:"reconcile"`
+}
```
Move types to the head of the file (directly after imports)

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+}
+
+func getConnectionString(cfg *config) string {
+	return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", cfg.DB.Host, cfg.DB.Port, cfg.DB.User, cfg.DB.Password, cfg.DB.Name)
```
break long lines
```suggestion
	return fmt.Sprintf(
		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable", 
		cfg.DB.Host, 
		cfg.DB.Port, 
		cfg.DB.User, 
		cfg.DB.Password, 
		cfg.DB.Name,
	)
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+	} `yaml:"reconcile"`
+}
+
+func getConnectionString(cfg *config) string {
```
The main module should not know about the internals of the db connection or its connection string format. Please move this into the db package

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+	pool      *pgxpool.Pool
+}
+
+func NewClient(ctx context.Context, connString, providerTableName string) (*Database, error) {
```
Don't assume the client only work on a single table, accept a map and validate that the input (map) contains a name mapping for all the tables you need in order to make the client work 
```suggestion
func NewClient(ctx context.Context, connString, tables map[string]string) (*Database, error) {
    if _, ok := tables["providers"]; ! ok {
        return fmt.Errorf("Incomplete table name mapping, missing a table name for \"providers\" table")
    }

```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+func NewClient(ctx context.Context, connString, providerTableName string) (*Database, error) {
+	pool, err := pgxpool.New(ctx, connString)
+	if err != nil {
+		return nil, err
```
Please wrap with an error that give better context

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	if err := pool.Ping(ctx); err != nil {
+		return nil, err
```
Please wrap with an error that give better context

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+	return &Database{
+		pool:      pool,
+		tableName: providerTableName,
+	}, nil
```
```suggestion
	db := Database{}
	db.pool = pool
	db.tables.providers = tables["providers"]
	return &db, nil
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+	pg.pool.Close()
+}
+
+func (pg *Database) ListenForNotifications(ctx context.Context, connString, providerTableName string, fn func(payload string)) error {
```
Why are we trying to implement a generic listen? I would prefer to be specific to the event/notification the we are trying to listen to.

Also, no need to provide the connection string or the providersTableName as you already have access to the Database object

```suggestion
func (pg *Database) OnProvider(ctx context.Context, fn func(providerName string)) error {
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	return nil
+}
```
No need for this entire function. Just add a boolean parameter to `OnProvider`, something like:
```notifyOnExisting bool```

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,109 @@
+	logger *zap.Logger,
+	client *db.Database,
+	workerCount int,
+	reconcile func(*zap.Logger, *db.Database, Request) Result,
```
```suggestion
	reconcile func(Request) Result,
```

nb-ohad commented on pkg/forman/forman.go:
```diff
@@ -0,0 +1,109 @@
+		for i := 0; i < workerCount; i++ {
+			workers.GoWorker(toWorkers, results, func(req Request) Result {
+				return reconcile(logger, client, req)
+			})
```
```suggestion
			workers.GoWorker(toWorkers, results, reconcile)
```

nb-ohad commented on pkg/reconciler/reconcile.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+func Reconcile(logger *zap.Logger, client *db.Database, req forman.Request) forman.Result {
+	ctx := context.Background()
```
Is having a new context for each reconcile by design?

Madhu-1 replied:
i was thinking of keeping new context per reconcile as it will be specific to a single provider, if you think we can have only one context I can remove this one.

nb-ohad replied:
No, the reasoning sounds good to me. Let's keep it with a separate context for each reconcile.

nb-ohad commented on pkg/types/types.go:
```diff
@@ -0,0 +1,17 @@
+}
+
+type Status struct {
+}
```
```suggestion
type ProviderClusterMetadata struct {
}
type ProviderClusterSpec struct {
}

type ProviderClusterStatus struct {
}
```

nb-ohad commented on pkg/types/types.go:
```diff
@@ -0,0 +1,17 @@
+	SatelliteID string
+	MetaData    Metadata
+	Spec        Spec
+	Status      Status
```
```suggestion
	MetaData    ProviderClusterMetadata
	Spec        ProviderClusterSpec
	Status      ProviderClusterStatus
```

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,121 @@
+		}); err != nil {
+			logger.Fatal(err.Error())
+		}
+	}()
```
This entire code is unnecessary, please refer to the comment I left about this function on the db package  
```suggestion
```

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func (pg *Database) OnProvider(ctx context.Context, logger *zap.Logger, notifyExisting bool, fn func(providerName string)) error {
+	if notifyExisting {
+		go func() {
+			query := fmt.Sprintf("SELECT * FROM %s", pg.tables.providers)
+			rows, err := pg.pool.Query(ctx, query)
+			if err != nil {
+				logger.Fatal("failed to query database", zap.Error(err))
+			}
+			defer rows.Close()
+
+			for rows.Next() {
+				var c types.ProviderCluster
+				if err := rows.Scan(&c.ClusterID, &c.AccountID, &c.SatelliteID, &c.MetaData, &c.Spec, &c.Status); err != nil {
+					logger.Fatal("failed to scan row", zap.Error(err))
+				}
+				fn(c.ClusterID)
+			}
+		}()
+	}
```
Running in a go routine can create a race condition between events based on existing records in the DB and new even coming from the listen. In theory, we have Forman taking care of parallel events on the same provider record, but I think we should not rely on other packages to solve problems that we can easily solve within this package. 

In addition, OnProvider will not return if it gets to the WaitForNotification part of the function (below). This means control will never go back to the caller func, which is not something I expect from a function that is supposed to set up a handler for an event/notification.  

I will suggestion to have the entire logic invoked in a single Go routine (from the start till the end) and not separate the first part into its own Go routine to make sure we preserve order.

@Madhu-1 WDYT?

Madhu-1 replied:
The reason for handling the already existing data in database and new events separately for below reason

* From my understanding, we scan all the entries from the database and process them serially we might miss below events
* If there is an update for the already processed provider
* If there is a new provider added

The above are the reason to start a listener and consider handling the objects in the database in a goroutine and as you mentioned depend on the forman to take care of for not for parallel processing a single provider.

Let me know what do you think, if we can discard above 2 cases I can have a single goroutine for this one.

nb-ohad replied:
Just to understand, you concern is that between the time we are reading the records from the DB and the time we start listening to notifications there might be a new provider record added to the DB?

Madhu-1 replied:
Yes there might be a new provider or any update on the already processed provider

Madhu-1 replied:
@nb-ohad as disucussed move the code to start listening for the update and process the existing data in goroutines and also added signal handling to close open connections.

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,115 @@
+	)
+
+	if err := dbClient.OnProvider(ctx, logger, true, func(payload string) {
+		req := forman.Request{}
+		if err := json.Unmarshal([]byte(payload), &req); err != nil {
+			return
+		}
+		reqChan <- req
```
I believe that with the current implementation, the payload is already the provider name and that the Unmarshal is unnecessary

Madhu-1 replied:
Yes correct, i will remove the json unmarshal

nb-ohad commented on main.go:
```diff
@@ -1,5 +1,123 @@
+	<-sigChan
+	logger.Info("Received signal, shutting down")
+	dbClient.Close(ctx)
+	os.Exit(0)
```
Why do we need these 2 lines? if we just let the function exist then you get the `dbClient.close()` automatically from the  `defer dbClient.Close(ctx)` and exist the program as this is the main go routine

Madhu-1 replied:
yes that correct, removed extra 2 lines

nb-ohad commented on pkg/db/db.go:
```diff
@@ -0,0 +1,118 @@
+	if _, err = pg.conn.Exec(ctx, query); err != nil {
+		return err
+	}
+	go func() {
+		for {
+			notification, err := pg.conn.Conn().WaitForNotification(ctx)
+			if err != nil {
+				// If the connection is closed, release the connection and exit the goroutine
+				if pg.conn.Conn().IsClosed() {
+					pg.conn.Release()
+					logger.Error("Connection closed by server:", zap.Error(err))
+					return
+				}
+				logger.Error("failed to wait for notification", zap.Error(err))
+				continue
+			}
+			fn(notification.Payload)
+		}
+	}()
+
+	if notifyExisting {
+		go func() {
+			query := fmt.Sprintf("SELECT cluster_id FROM %s", pg.tables.providers)
+			rows, err := pg.pool.Query(ctx, query)
+			if err != nil {
+				logger.Fatal("failed to query database", zap.Error(err))
+			}
+			defer rows.Close()
+
+			for rows.Next() {
+				var clusterID string
+				if err := rows.Scan(&clusterID); err != nil {
+					logger.Fatal("failed to scan row", zap.Error(err))
+				}
+				fn(clusterID)
+			}
+		}()
+	}
```
I am not sure this addresses the concern raised here: https://github.com/red-hat-storage/managed-fusion-fleet-reconciler/pull/5#discussion_r1266730976

Mainly as you are just invoking 2 go routines one after another, but I am not sure go has any kind of guarantees that the second go routine will start execution after the first one. This might result in very subtle bugs. 

If go has a guarantee in place (somewhere in the language spec) then this code is ok

Madhu-1 replied:
Yes, we are invoking 2 go routines one after the other after subscribing for the notification it doesn't matter which go routine runs first, one goroutine read from the notification and another reads from the database and processes the data. The point to note here is that we are not subscribing to the events in the goroutine it's done before starting the goroutines,  AFAIK it addresses the concern raised above. Let me know if it makes sense or if we need to change anything here.

nb-ohad replied:
Ok, I understand now. Thanks for the clarification!

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+type Database struct {
+	tables struct {
+		providers string
+	}
```
- is there any reason why this can't be a map?

Madhu-1 replied:
This is a suggestion made by @nb-ohad. @nb-ohad can help you here

nb-ohad replied:
Why do we need a map (which is not type-safe) if we know exactly what the keys are?

leelavg replied:
- pardon I didn't understand the premise
- Don't we want to just store keys as short form for table names and the type is always a string, isn't it?

nb-ohad replied:
Let me put it a little bit differently. The DB lib knows exactly which tables it supports and which columns it expects.
This is not a generic lib to handle a database, if we wanted a generic lib I will argue the underlying lib is enough.

As such there is no reason to introduce keys as strings, mainly as this might result in a mistype somewhere and a subtle bug. Accessing `a.b` had less error surface than accessing using `a["b"]`? In the first case, if you mistype b you will get a compile-time error while in the second case you might get a runtime error if you have put some code to check for it.

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,118 @@
+
+	if _, ok := tables["providers"]; !ok {
+		return nil, fmt.Errorf("incomplete table name mapping, missing a table name for \"providers\" table")
+	}
```
- can this be moved to start of the function as it is a standalone check?

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	query := fmt.Sprintf("LISTEN %s;", pg.tables.providers)
+	// Listen for notifications
```
- if you don't mind, pls change this comment, currently it reads as a blocking call which actually it isn't

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,118 @@
+	}
+	go func() {
+		for {
+			notification, err := pg.conn.Conn().WaitForNotification(ctx)
```
- just curious, were you able to test with providers count more than workers with no worker is available for work?
- I know in the prototype we didn't use db notifications, simply wanted to know if these gets buffered or not, a link to the doc also suffices.

Madhu-1 replied:
This is not tested with DB i don't see a problem with that, if you see a problem I can take a look at it now or if we get a problem in this area it can be addressed later.

leelavg reviewed (commented):
- thx for all the revisions, this PR shaped quite nicely :)

leelavg commented on pkg/db/db.go:
```diff
@@ -0,0 +1,117 @@
+		}
+	}()
+
+	if notifyExisting {
```
- I believe this'll be the last comment from me, I don't foresee any scenario where this'll be false, do you see any?

Madhu-1 replied:
This might get change in the future for now, as @nb-ohad thinks we might have a scenario like that going forward we might need to keep it and remove it later if not at all required.

nb-ohad replied:
@leelavg you are raising a good point. Right now nothing prevents us from calling OnProvider more than once. my assumption was that when we lose the connection (for one reason or another) we will want to error handle and reestablish the notification channel without re-running the reconcile for all records.

leelavg replied:
Ack, I was thinking more like to reconcile all records with duration let's say every 10 min and accepting duration seems more suitable.

Nevertheless, this isn't a blocker for now.
